cosmwasm/packages/wasmi-runtime/src/wasm/query_chain.rs

```diff
@@ -0,0 +1,232 @@
+use super::errors::WasmEngineError;
+use crate::crypto::Ed25519PublicKey;
+use crate::wasm::types::{IoNonce, SecretMessage};
+use crate::{exports, imports};
+
+use crate::cosmwasm::encoding::Binary;
+use crate::cosmwasm::query::{QueryRequest, WasmQuery};
+use crate::cosmwasm::{
+    std_error::{StdError, StdResult},
+    system_error::SystemResult,
+};
+
+use enclave_ffi_types::{Ctx, EnclaveBuffer, OcallReturn, UntrustedVmError};
+use log::*;
+use sgx_types::sgx_status_t;
+
+pub fn encrypt_and_query_chain(
+    query: &[u8],
+    context: &Ctx,
+    nonce: IoNonce,
+    user_public_key: Ed25519PublicKey,
+) -> Result<(Vec<u8>, u64), WasmEngineError> {
+    let mut query_struct: QueryRequest = serde_json::from_slice(query).map_err(|err| {
+        error!(
+            "encrypt_and_query_chain() cannot build struct from json {:?}: {:?}",
+            String::from_utf8_lossy(query),
+            err
+        );
+        WasmEngineError::BadQueryChainRequest
+    })?;
+
+    let mut is_encrypted = false;
+
+    if let QueryRequest::Wasm(WasmQuery::Smart {
+        contract_addr: _,
+        msg,
+    }) = &mut query_struct
+    {
+        is_encrypted = true;
+
+        let mut encrypted_msg = SecretMessage {
+            msg: msg.0.clone(),
+            user_public_key,
+            nonce,
+        };
+        encrypted_msg.encrypt_in_place().map_err(|err|{
+            error!(
+                "encrypt_and_query_chain() got an error while trying to encrypt the request for query {:?}, stopping wasm: {:?}",
+                String::from_utf8_lossy(&query),
+                err
+            );
+
+            WasmEngineError::EncryptionError
+        })?;
+
+        *msg = Binary(encrypted_msg.to_slice());
+    }
+
+    let encrypted_query = serde_json::to_vec(&query_struct).map_err(|err| {
+        // this should never happen
+        error!(
+            "encrypt_and_query_chain() cannot build json from struct {:?}: {:?}",
+            query_struct, err
+        );
+        WasmEngineError::BadQueryChainRequest
+    })?;
+
+    let answer = query_chain(context, &encrypted_query);
+
+    if !is_encrypted {
+        return answer;
+    }
+
+    // Call query_chain (this bubbles up to x/compute via ocalls and FFI to Go code)
+    // This returns the answer from x/compute
+    match answer {
+        Ok((encrypted_answer_as_vec, gas_used)) => {
+            // answer is QueryResult (Result<Binary,StdError>) encoded be serde to bytes
```

Comment:
This match can be replaced by:
```rust
let (encrypted_answer_as_vec, gas_used) = answer?;
// The main arm of the match here.
```
this will also help flatten the code (but not The Curve, unfortunately)